Wip/devel/slurm bug 17399 integ test

jdeamicis opened the pull request:
### Description of changes
* Add integration test to test Slurm bug 17399.
* Add support for the `--test-only` sbatch option in the `SlurmCommands` class.
* Minor improvements to the `test_slurm` test.

### Tests
* Manually ran the new test and the modified `test_slurm`.

### Checklist
- Make sure you are pointing to **the right branch**.
- If you're creating a patch for a branch other than `develop` add the branch name as prefix in the PR title (e\.g\. `[release-3.6]`).
- Check all commits' messages are clear, describing what and why vs how.
- Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jdeamicis committed:
Extend Slurm submission commands for `--test-only`

Add new parameter for Slurm submission methods to support the `--test-only`
option of `sbatch`, which allows to test the submission of jobs without
actually submitting them.

Adapt existing integration tests that were using the `--test-only` inside
the `other_options` parameter.

Adapt abstract method signature for assert_job_submitted

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Improve test_slurm

Remove confusing references to `scheduler` variable (we are testing
slurm here, no other scheduler would make sense).

Get GPU hardware information from the instance type data instead of
hard-coding them in the test.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add test_slurm_ticket_17399

Add integration test to test issues in Slurm job submission
due to ticket 17399.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

NSsirena commented on tests/integration-tests/tests/schedulers/test_slurm.py:
```diff
@@ -113,25 +113,70 @@ def test_slurm(
+    disable_multithreading = False
+    gpu_instance_type_info = get_instance_info(gpu_instance_type, region)
+    gpus_per_instance = _get_num_gpus_on_instance(gpu_instance_type_info)
+    if disable_multithreading:
+        cpus_per_instance = gpu_instance_type_info.get("VCpuInfo").get("DefaultCores")
+    else:
```
We can remove these since we are forcing `disable_multithreading = False` few lines above

jdeamicis replied:
I wanted that `disable_multithreading` to be used as a flag, but since it's not used anywhere I will remove it.

jdeamicis replied:
Addressed.

NSsirena commented on tests/integration-tests/tests/schedulers/test_slurm.py:
```diff
@@ -113,25 +113,70 @@ def test_slurm(
+def test_slurm_bug_17399(
+    region, pcluster_config_reader, clusters_factory, test_datadir, architecture, scheduler_commands_factory
+):
+    """
+    Test Slurm bug 17399
+
+    Check if certain valid combinations of sbatch options lead to a submission error
```
Please add some extra details on what this code will actually test.

While `bug 17399` is really specific ...it is hard to decode for anyone that wasn't involved in the investigation ...and I doubt we will remember the specific case in few months.

jdeamicis replied:
We discussed offline and decided just to rename the test.

NSsirena commented on tests/integration-tests/configs/schedulers.yaml:
```diff
@@ -136,6 +136,12 @@ test-suites:
           instances: ["c5.xlarge"]
           oss: ["alinux2"]
           schedulers: ["slurm"]
+    test_slurm.py::test_slurm_ticket_17399:
```
Since you changed the name of the test you should also change the name of the subfolder with the config file, otherwise the test will fail

jdeamicis replied:
I had done it yesterday, but I forgot to force-push.

NSsirena commented on tests/integration-tests/tests/common/schedulers_common.py:
```diff
@@ -251,9 +251,12 @@ def get_job_submit_time(self, job_id):  # noqa: D102
         return self.get_job_info(job_id, field="EligibleTime")
 
-    def assert_job_submitted(self, sbatch_output):  # noqa: D102
+    def assert_job_submitted(self, sbatch_output, test_only: bool = False):  # noqa: D102
```
See CodeQL warning: you'd have to add this parameter also in the abstract method.

jdeamicis committed:
Address code linters and review comments

Add noqa: C901 for _submit_batch_job method to avoid triggering
the rule due to the many if clauses.

Remove useless logic in the new test to define cpus_per_instance.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>
